File: cli/src/runtime/external/gateway-task.ts
import fs from "node:fs"
import os from "node:os"
import path from "node:path"
import type { NodeExternalAgentSpawnConfig } from "./node-backend"

const PAYLOAD_ENV = "COGNIA_GATEWAY_TASK_CONFIG"
const ALLOWED_FILES = new Set([
  "codex/config.toml",
  "pi/models.json",
  "pi/settings.json",
  "qwen/settings.json",
])
const ESSENTIAL_ENV = new Set([
  "PATH",
  "LANG",
  "LC_ALL",
  "LC_CTYPE",
  "TERM",
  "COLORTERM",
  "TMPDIR",
  "TMP",
  "TEMP",
  "SYSTEMROOT",
  "WINDIR",
  "SSL_CERT_FILE",
  "SSL_CERT_DIR",
  "NODE_EXTRA_CA_CERTS",
])

export function gatewayRuntimeEnvironment(ambient: NodeJS.ProcessEnv): NodeJS.ProcessEnv {
  return Object.fromEntries(Object.entries(ambient).filter(([key]) => ESSENTIAL_ENV.has(key)))
}

export function deleteGatewayTask(taskId: string, home = os.homedir()): void {
  if (!/^[a-zA-Z0-9_-]{1,128}$/.test(taskId)) throw new Error("Invalid gateway task id")
  const parent = path.join(home, ".local/share/cognia-agent-tasks")
  const root = path.join(parent, taskId)
  for (const directory of [parent, root]) {
    if (fs.existsSync(directory) && fs.lstatSync(directory).isSymbolicLink())
      throw new Error("Gateway task state must not be a symlink")
  }
  fs.rmSync(root, { recursive: true, force: true })
}

function privateDirectory(directory: string): void {
  if (fs.existsSync(directory) && fs.lstatSync(directory).isSymbolicLink())
    throw new Error("Gateway task state must not be a symlink")
  if (!fs.existsSync(path.dirname(directory))) privateDirectory(path.dirname(directory))
  fs.mkdirSync(directory, { recursive: true, mode: 0o700 })
  fs.chmodSync(directory, 0o700)
}

function privateWrite(filename: string, contents: string): void {
  if (fs.existsSync(filename) && fs.lstatSync(filename).isSymbolicLink())
    throw new Error("Gateway task file must not be a symlink")
  privateDirectory(path.dirname(filename))
  fs.writeFileSync(filename, contents, { mode: 0o600 })
  fs.chmodSync(filename, 0o600)
}

/** Fixed host-owned paths retain conversation state; config files never contain the lease. */
export function prepareGatewayTask(
  config: NodeExternalAgentSpawnConfig,
  home = os.homedir()
): {
  config: NodeExternalAgentSpawnConfig
  cleanup: () => void
} {
  const raw = config.env?.[PAYLOAD_ENV]
  if (!raw) return { config, cleanup() {} }
  const payload = JSON.parse(raw) as {
    taskId: string
    binding: unknown
    runtime: string
    ownerAccountId: string | null
    files: Record<string, string>
  }
  if (
    !/^[a-zA-Z0-9_-]{1,128}$/.test(payload.taskId) ||
    !["codex", "opencode", "pi", "claude", "qwen"].includes(payload.runtime) ||
    !payload.files ||
    Object.entries(payload.files).some(
      ([name, value]) =>
        !ALLOWED_FILES.has(name) || typeof value !== "string" || value.length > 262144
    )
  ) {
    throw new Error("Invalid gateway task configuration")
  }
  let root = path.join(home, ".local/share/cognia-agent-tasks", payload.taskId)
  privateDirectory(root)
  root = fs.realpathSync(root)
  const bindingPath = path.join(root, "binding.json")
  const binding = JSON.stringify({
    binding: payload.binding,
    runtime: payload.runtime,
    ownerAccountId: payload.ownerAccountId ?? null,
  })
  if (fs.existsSync(bindingPath)) {
    if (fs.readFileSync(bindingPath, "utf8") !== binding)
      throw new Error("This task is bound to a different model or account; start a new task")
  } else privateWrite(bindingPath, binding)
  const files: string[] = []
  const cleanup = () => {
    for (const filename of files) fs.rmSync(filename, { force: true })
  }
  try {
    for (const [name, contents] of Object.entries(payload.files)) {
      const filename = path.join(root, name)
      privateWrite(filename, contents)
      files.push(filename)
    }
    const env = { ...config.env }
    delete env[PAYLOAD_ENV]
    env.COGNIA_GATEWAY_TASK_HOME = root
    for (const [key, relative] of Object.entries({
      HOME: "",
      USERPROFILE: "",
      XDG_CONFIG_HOME: "config",
      XDG_DATA_HOME: "data",
      XDG_CACHE_HOME: "cache",
      XDG_STATE_HOME: "state",
      CODEX_HOME: "codex",
      PI_CODING_AGENT_DIR: "pi",
      CLAUDE_CONFIG_DIR: "claude",
      OPENCODE_CONFIG_DIR: "config/opencode",
    })) {
      const directory = path.join(root, relative)
      privateDirectory(directory)
      env[key] = directory
    }
    if (payload.runtime === "qwen") {
      env.QWEN_HOME = path.join(root, "qwen")
      env.QWEN_RUNTIME_DIR = path.join(root, "qwen-runtime")
      privateDirectory(env.QWEN_RUNTIME_DIR)
      env.QWEN_CODE_SYSTEM_SETTINGS_PATH = path.join(root, "qwen/settings.json")
      env.QWEN_CODE_SYSTEM_DEFAULTS_PATH = env.QWEN_CODE_SYSTEM_SETTINGS_PATH
    }
    return { config: { ...config, env }, cleanup }
  } catch (error) {
    cleanup()
    throw error
  }
}

File: crates/cognia-external-agent/src/gateway_task.rs
//! Host-owned task state. Provider files contain environment references, never lease secrets.
use std::{collections::HashMap, fs, path::{Path, PathBuf}};
use serde::Deserialize;

pub const PAYLOAD_ENV: &str = "COGNIA_GATEWAY_TASK_CONFIG";

#[derive(Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct TaskPayload {
    task_id: String,
    binding: serde_json::Value,
    owner_account_id: Option<String>,
    runtime: String,
    files: HashMap<String, String>,
}

fn payload(env: &HashMap<String, String>) -> Result<Option<TaskPayload>, String> {
    let Some(raw) = env.get(PAYLOAD_ENV) else { return Ok(None) };
    let value: TaskPayload = serde_json::from_str(raw).map_err(|_| "Invalid gateway task configuration")?;
    if value.task_id.is_empty() || value.task_id.len() > 128 ||
        !value.task_id.bytes().all(|c| c.is_ascii_alphanumeric() || c == b'-' || c == b'_') ||
        !["codex", "opencode", "pi", "claude", "qwen", "dsh"].contains(&value.runtime.as_str()) ||
        value.files.iter().any(|(name, contents)| !["codex/config.toml", "pi/models.json", "pi/settings.json", "qwen/settings.json"].contains(&name.as_str()) || contents.len() > 262144) {
        return Err("Invalid gateway task configuration".into());
    }
    Ok(Some(value))
}

pub fn task_home(env: &HashMap<String, String>, home: &Path) -> Result<Option<PathBuf>, String> {
    let home = home.canonicalize().unwrap_or_else(|_| home.to_path_buf());
    Ok(payload(env)?.map(|value| home.join(".local/share/cognia-agent-tasks").join(value.task_id)))
}

fn private_dir(path: &Path) -> Result<(), String> {
    if fs::symlink_metadata(path).is_ok_and(|meta| meta.file_type().is_symlink()) {
        return Err("Gateway task state must not be a symlink".into());
    }
    if let Some(parent) = path.parent() {
        if !parent.exists() { private_dir(parent)?; }
    }
    fs::create_dir_all(path).map_err(|e| format!("Cannot create gateway task state: {e}"))?;
    #[cfg(unix)] {
        use std::os::unix::fs::PermissionsExt;
        fs::set_permissions(path, fs::Permissions::from_mode(0o700)).map_err(|e| e.to_string())?;
    }
    Ok(())
}

fn write_private(path: &Path, contents: &[u8]) -> Result<(), String> {
    if fs::symlink_metadata(path).is_ok_and(|meta| meta.file_type().is_symlink()) {
        return Err("Gateway task file must not be a symlink".into());
    }
    private_dir(path.parent().ok_or("Invalid gateway task path")?)?;
    fs::write(path, contents).map_err(|e| format!("Cannot write gateway task configuration: {e}"))?;
    #[cfg(unix)] {
        use std::os::unix::fs::PermissionsExt;
        fs::set_permissions(path, fs::Permissions::from_mode(0o600)).map_err(|e| e.to_string())?;
    }
    Ok(())
}

pub struct TaskFiles { files: Vec<PathBuf> }
impl Drop for TaskFiles {
    fn drop(&mut self) {
        for file in &self.files { let _ = fs::remove_file(file); }
    }
}

pub fn prepare(env: &mut HashMap<String, String>, home: &Path) -> Result<Option<TaskFiles>, String> {
    let Some(value) = payload(env)? else { return Ok(None) };
    let root = task_home(env, home)?.ok_or("Missing gateway task home")?;
    private_dir(&root)?;
    let binding_path = root.join("binding.json");
    let binding = serde_json::json!({ "binding": value.binding, "runtime": value.runtime, "ownerAccountId": value.owner_account_id });
    if binding_path.exists() {
        let existing: serde_json::Value = serde_json::from_slice(&fs::read(&binding_path).map_err(|e| e.to_string())?).map_err(|_| "Invalid saved gateway task binding")?;
        if existing != binding { return Err("This task is bound to a different model or account; start a new task".into()); }
    } else {
        write_private(&binding_path, &serde_json::to_vec(&binding).map_err(|e| e.to_string())?)?;
    }
    let mut guard = TaskFiles { files: vec![] };
    for (name, contents) in value.files {
        let path = root.join(name);
        write_private(&path, contents.as_bytes())?;
        guard.files.push(path);
    }
    for (key, relative) in [("HOME", ""), ("USERPROFILE", ""), ("XDG_CONFIG_HOME", "config"), ("XDG_DATA_HOME", "data"), ("XDG_CACHE_HOME", "cache"), ("XDG_STATE_HOME", "state"), ("CODEX_HOME", "codex"), ("PI_CODING_AGENT_DIR", "pi"), ("CLAUDE_CONFIG_DIR", "claude"), ("OPENCODE_CONFIG_DIR", "config/opencode")] {
        let path = if relative.is_empty() { root.clone() } else { root.join(relative) };
        private_dir(&path)?;
        env.insert(key.into(), path.to_string_lossy().into_owned());
    }
    if value.runtime == "qwen" {
        let runtime = root.join("qwen-runtime");
        private_dir(&runtime)?;
        env.insert("QWEN_HOME".into(), root.join("qwen").to_string_lossy().into_owned());
        env.insert("QWEN_RUNTIME_DIR".into(), runtime.to_string_lossy().into_owned());
        for key in ["QWEN_CODE_SYSTEM_SETTINGS_PATH", "QWEN_CODE_SYSTEM_DEFAULTS_PATH"] {
            env.insert(key.into(), root.join("qwen/settings.json").to_string_lossy().into_owned());
        }
    }
    env.remove(PAYLOAD_ENV);
    Ok(Some(guard))
}

/// Runtime essentials only. No auth, provider settings, home, or shell injection.
pub fn runtime_environment() -> HashMap<String, String> {
    ["PATH", "LANG", "LC_ALL", "LC_CTYPE", "TERM", "COLORTERM", "TMPDIR", "TMP", "TEMP", "SYSTEMROOT", "WINDIR", "SSL_CERT_FILE", "SSL_CERT_DIR", "NODE_EXTRA_CA_CERTS"]
        .into_iter().filter_map(|key| std::env::var(key).ok().map(|value| (key.to_string(), value))).collect()
}

pub fn delete_task(task_id: &str, home: &Path) -> Result<(), String> {
    if task_id.is_empty() || task_id.len() > 128 || !task_id.bytes().all(|c| c.is_ascii_alphanumeric() || c == b'-' || c == b'_') {
        return Err("Invalid gateway task id".into());
    }
    let parent = home.join(".local/share/cognia-agent-tasks");
    if !parent.exists() { return Ok(()) }
    if fs::symlink_metadata(&parent).is_ok_and(|meta| meta.file_type().is_symlink()) {
        return Err("Gateway task state must not be a symlink".into());
    }
    let root = parent.join(task_id);
    if !root.exists() { return Ok(()) }
    if fs::symlink_metadata(&root).is_ok_and(|meta| meta.file_type().is_symlink()) {
        return Err("Gateway task state must not be a symlink".into());
    }
    fs::remove_dir_all(root).map_err(|e| format!("Cannot remove gateway task state: {e}"))
}

#[cfg(test)]
mod tests {
    use super::*;
    fn env(task: &str) -> HashMap<String, String> {
        HashMap::from([(PAYLOAD_ENV.into(), serde_json::json!({"taskId":task,"binding":{"providerId":"gateway","modelId":"model"},"runtime":"pi","files":{"pi/models.json":"{}"}}).to_string())])
    }
    #[test]
    fn isolates_configuration_preserves_history_and_refuses_rebinding() {
        let dir = tempfile::tempdir().unwrap();
        let mut input = env("one");
        let root = task_home(&input, dir.path()).unwrap().unwrap();
        let guard = prepare(&mut input, dir.path()).unwrap();
        assert_eq!(input["HOME"], root.to_string_lossy());
        assert!(!input.contains_key(PAYLOAD_ENV));
        fs::write(root.join("pi/session.jsonl"), "conversation").unwrap();
        drop(guard);
        assert!(!root.join("pi/models.json").exists());
        assert!(root.join("pi/session.jsonl").exists());
        let mut changed = env("one");
        changed.get_mut(PAYLOAD_ENV).unwrap().push(' ');
        assert!(prepare(&mut changed, dir.path()).is_ok());
        let mut changed = env("one");
        changed.insert(PAYLOAD_ENV.into(), changed[PAYLOAD_ENV].replace("\"model\"", "\"other\""));
        assert!(prepare(&mut changed, dir.path()).is_err());
    }
    #[test]
    fn rejects_path_traversal_and_unknown_files() {
        let dir = tempfile::tempdir().unwrap();
        assert!(prepare(&mut env("../escape"), dir.path()).is_err());
        let mut input = env("ok");
        input.insert(PAYLOAD_ENV.into(), input[PAYLOAD_ENV].replace("pi/models.json", "../secret"));
        assert!(prepare(&mut input, dir.path()).is_err());
    }

    #[test]
    fn pins_qwen_system_settings_and_retains_session_state() {
        let dir = tempfile::tempdir().unwrap();
        let mut input = env("qwen-one");
        input.insert(PAYLOAD_ENV.into(), input[PAYLOAD_ENV].replace("\"pi\"", "\"qwen\"").replace("pi/models.json", "qwen/settings.json"));
        let original = input.clone();
        let guard = prepare(&mut input, dir.path()).unwrap();
        let settings = PathBuf::from(&input["QWEN_CODE_SYSTEM_SETTINGS_PATH"]);
        assert_eq!(settings, PathBuf::from(&input["QWEN_HOME"]).join("settings.json"));
        assert_eq!(input["QWEN_CODE_SYSTEM_DEFAULTS_PATH"], input["QWEN_CODE_SYSTEM_SETTINGS_PATH"]);
        let history = PathBuf::from(&input["QWEN_RUNTIME_DIR"]).join("session.jsonl");
        fs::write(&history, "history").unwrap();
        drop(guard);
        assert!(!settings.exists());
        let mut resumed = original;
        let guard = prepare(&mut resumed, dir.path()).unwrap();
        assert_eq!(fs::read_to_string(&history).unwrap(), "history");
        drop(guard);
    }

    #[test]
    fn dsh_gateway_preserves_certified_state_and_keeps_lease_out_of_files() {
        let dir = tempfile::tempdir().unwrap();
        let mut input = env("dsh-one");
        input.insert(PAYLOAD_ENV.into(), serde_json::json!({"taskId":"dsh-one","binding":{"providerId":"gateway","modelId":"model"},"runtime":"dsh","files":{}}).to_string());
        input.insert("DSH_HOME".into(), "/managed/dsh/dsh-home".into());
        input.insert("COGNIA_DSH_SESSION_ROOT".into(), "/managed/dsh/sessions".into());
        input.insert("COGNIA_DSH_GATEWAY_TOKEN".into(), "temporary-lease".into());
        input.insert("COGNIA_DSH_GATEWAY_CONFIG".into(), "{\"providers\":{}}".into());
        let root = task_home(&input, dir.path()).unwrap().unwrap();
        let guard = prepare(&mut input, dir.path()).unwrap();
        assert_eq!(input["DSH_HOME"], "/managed/dsh/dsh-home");
        assert_eq!(input["COGNIA_DSH_SESSION_ROOT"], "/managed/dsh/sessions");
        assert_eq!(input["COGNIA_DSH_GATEWAY_TOKEN"], "temporary-lease");
        assert!(!input.contains_key(PAYLOAD_ENV));
        assert!(!fs::read_to_string(root.join("binding.json")).unwrap().contains("temporary-lease"));
        drop(guard);
        delete_task("dsh-one", dir.path()).unwrap();
        assert!(!root.exists());
    }

    #[test]
    fn freezes_local_owner_and_deletes_only_the_selected_task() {
        let dir = tempfile::tempdir().unwrap();
        let mut input = env("one");
        let raw = input.get_mut(PAYLOAD_ENV).unwrap();
        *raw = raw.replacen('{', "{\"ownerAccountId\":\"owner-one\",", 1);
        let guard = prepare(&mut input, dir.path()).unwrap();
        drop(guard);
        let mut other = env("one");
        let raw = other.get_mut(PAYLOAD_ENV).unwrap();
        *raw = raw.replacen('{', "{\"ownerAccountId\":\"owner-two\",", 1);
        assert!(prepare(&mut other, dir.path()).is_err());
        let mut sibling = env("two");
        let sibling_guard = prepare(&mut sibling, dir.path()).unwrap();
        delete_task("one", dir.path()).unwrap();
        assert!(!task_home(&env("one"), dir.path()).unwrap().unwrap().exists());
        assert!(task_home(&env("two"), dir.path()).unwrap().unwrap().exists());
        drop(sibling_guard);
    }
}
